File: frontend/src/api/workflows.ts
import { apiRequest } from "./client";
import type { ConfigPreview, WorkflowDraft, WorkflowDraftRequest } from "./types";

export function createWorkflowDraft(request: WorkflowDraftRequest): Promise<WorkflowDraft> {
  return apiRequest<WorkflowDraft>("/api/workflows", {
    method: "POST",
    body: JSON.stringify(request),
  });
}

export function getWorkflowDraft(workflowId: string): Promise<WorkflowDraft> {
  return apiRequest<WorkflowDraft>(`/api/workflows/${workflowId}`);
}

export function patchWorkflowDraft(workflowId: string, request: WorkflowDraftRequest): Promise<WorkflowDraft> {
  return apiRequest<WorkflowDraft>(`/api/workflows/${workflowId}`, {
    method: "PATCH",
    body: JSON.stringify(request),
  });
}

export function previewWorkflowConfig(workflowId: string): Promise<ConfigPreview> {
  return apiRequest<ConfigPreview>(`/api/workflows/${workflowId}/config/preview`, {
    method: "POST",
  });
}

File: backend/app/api/workflows.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException

from backend.app.schemas.workflow import ConfigPreview, WorkflowCreateRequest, WorkflowDraft, WorkflowPatchRequest
from backend.app.services.config_builder import build_config_preview
from backend.app.services.workflow_store import create_workflow, get_workflow, patch_workflow

router = APIRouter(tags=["workflows"])


@router.post("/workflows", response_model=WorkflowDraft)
def create_workflow_draft(request: WorkflowCreateRequest) -> WorkflowDraft:
    return create_workflow(request)


@router.get("/workflows/{workflow_id}", response_model=WorkflowDraft)
def read_workflow_draft(workflow_id: str) -> WorkflowDraft:
    try:
        return get_workflow(workflow_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.patch("/workflows/{workflow_id}", response_model=WorkflowDraft)
def update_workflow_draft(workflow_id: str, request: WorkflowPatchRequest) -> WorkflowDraft:
    try:
        return patch_workflow(workflow_id, request)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.post("/workflows/{workflow_id}/config/preview", response_model=ConfigPreview)
def preview_workflow_config(workflow_id: str) -> ConfigPreview:
    try:
        draft = get_workflow(workflow_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return build_config_preview(draft)
